sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala

```diff
@@ -0,0 +1,316 @@
+   * distinct count of other attributes based on output rows after join.
+   */
+  private def updateAttrStats(
+      outputRows: BigInt,
+      attributes: Seq[Attribute],
+      oldAttrStats: AttributeMap[ColumnStat],
+      joinKeyStats: AttributeMap[ColumnStat]): Seq[(Attribute, ColumnStat)] = {
+    val outputAttrStats = new ArrayBuffer[(Attribute, ColumnStat)]()
+    val leftRows = leftStats.rowCount.get
+    val rightRows = rightStats.rowCount.get
+    if (outputRows == leftRows * rightRows) {
+      // Cartesian product, just propagate the original column stats
+      attributes.foreach(a => outputAttrStats += a -> oldAttrStats(a))
+    } else if (outputRows != 0) {
+      val leftRatio =
+        if (leftRows != 0) BigDecimal(outputRows) / BigDecimal(leftRows) else BigDecimal(0)
+      val rightRatio =
+        if (rightRows != 0) BigDecimal(outputRows) / BigDecimal(rightRows) else BigDecimal(0)
+      attributes.foreach { a =>
+        // check if this attribute is a join key
+        if (joinKeyStats.contains(a)) {
+          outputAttrStats += a -> joinKeyStats(a)
+        } else {
+          val oldCS = oldAttrStats(a)
+          val oldNdv = oldCS.distinctCount
+          // We only change (scale down) the number of distinct values if the number of rows
+          // decreases after join, because join won't produce new values even if the number of
+          // rows increases.
+          val newNdv = if (join.left.outputSet.contains(a) && leftRatio < 1) {
+            ceil(BigDecimal(oldNdv) * leftRatio)
+          } else if (join.right.outputSet.contains(a) && rightRatio < 1) {
+            ceil(BigDecimal(oldNdv) * rightRatio)
+          } else {
+            oldNdv
+          }
+          // TODO: support nullCount updates for specific outer joins
+          outputAttrStats += a -> oldCS.copy(distinctCount = newNdv)
+        }
+      }
+    }
+    outputAttrStats
+  }
+
+  /** Get intersected column stats for join keys. */
+  private def getIntersectedStats(joinKeyPairs: Seq[(AttributeReference, AttributeReference)])
+    : AttributeMap[ColumnStat] = {
+
+    val intersectedStats = new mutable.HashMap[Attribute, ColumnStat]()
+    joinKeyPairs.foreach { case (leftKey, rightKey) =>
+      // Check if the two sides are disjoint
+      val leftKeyStats = leftStats.attributeStats(leftKey)
+      val rightKeyStats = rightStats.attributeStats(rightKey)
+      val lRange = Range(leftKeyStats.min, leftKeyStats.max, leftKey.dataType)
+      val rRange = Range(rightKeyStats.min, rightKeyStats.max, rightKey.dataType)
+      if (Range.isIntersected(lRange, rRange)) {
+        // Update intersected column stats
+        val minNdv = leftKeyStats.distinctCount.min(rightKeyStats.distinctCount)
+        val (newMin1, newMax1, newMin2, newMax2) =
+          Range.intersect(lRange, rRange, leftKey.dataType, rightKey.dataType)
+        intersectedStats.put(leftKey, intersectedColumnStat(leftKeyStats, minNdv,
```

Comment:
logically the join keys should have same column stats, we can write it more explicitly
```
assert(leftKey.dataType.sameType(rightKey.dataType))
val stats = ColumnStats(minNdv, newMin, newMax, nullCount = 0) // and some more logic to update the avg/max length.
intersectedStats.put(leftKey, stats)
intersectedStats.put(rightKey, stats)
```